### PR TITLE
Stacks Pay: A Payment Request Standard for Stacks Blockchain Payments

### PR DESCRIPTION
**Pull request by dantrevino**
Stacks Pay is a proposed standard for wrapping transaction information in a standard way for easy sharing.

### CODE REVIEW

**Comment by binxbrigandine** [comment]
Love standardization!

**Comment by markmhendrickson** [comment]
Thanks for proposing, @dantrevino! We'll review on the Leather team and provide thoughts.

**Review by kyranjamie — COMMENTED** [review]
Great work @dantrevino

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+- dueDate: MAY be included.
+
+- memo: MAY be included. MUST NOT include personal information.
+
+**Parameters to Ignore:**
+
+contractName, functionName: Wallets MUST ignore these parameters if present.
+
+### `mint`
```
[the remark this answers is not in the document]
Comment (dantrevino) [reply sips/sip-02x/sip-02x-stacks-pay.md:107]: I believe that enabling payments to smart contracts is within the realm of payments.  And malicious 'mint' operations can happen anywhere today.  At Boom we're putting default post conditions on every request.

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+
+**Stacks Pay** addresses this gap by introducing a payment request standard tailored specifically for the Stacks blockchain. The standard simplifies the process of sending and receiving payments, enhances interoperability among wallets and applications, and increases security by providing mechanisms to tie transactions back to payment requests.
+
+Stacks Pay does _not_ define smart contracts that may be used to facilitate specific functions, like subscriptions or NFT mints. Stacks Pay is strictly focused on defining the transaction payment information structure and encoding.
+
+## Operations
+
+Stacks Pay defines several standard operations that specify the type of payment action to be performed. Each operation type has specific required and optional parameters.
+
+### `support`
```
[the remark this answers is not in the document]
Comment (dantrevino) [reply sips/sip-02x/sip-02x-stacks-pay.md:51]: Open to other ideas here ... i tried to find a word that would encapsulate tips, donations, and gifts.

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+- functionName: A public contract function, typically `mint` or `claim`
+
+- token: MUST be either 'STX' or a valid SIP-010 contract name.
+
+- amount: The amount to be paid.
+
+**Optional Parameters:**
+
+- recipient: MAY be included; if present, MUST be a valid Stacks address. The NFT will be minted for the specified address.
```
Comment (whoabuddy) [inline sips/sip-02x/sip-02x-stacks-pay.md:125]: Where is it minted if this is not specified?

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+Applications **MAY** define custom operations for specific use cases. Custom operation types **MUST** be prefixed to prevent naming conflicts (e.g., `'custom-example'`).
+
+- **Handling Unrecognized Operations**: If a wallet encounters an unrecognized operation type, it **SHOULD**:
+
+- **Warn the User**: Inform the user that the operation type is unrecognized.
+
+- **Provide Safe Defaults**: Default to a standard payment flow if possible.
+
+- **Fail Gracefully**: Prevent unexpected behavior or security risks.
```
Comment (whoabuddy) [inline sips/sip-02x/sip-02x-stacks-pay.md:149]: Felt like this made it more readable

```suggestion
### Custom Operations

Applications **MAY** define custom operations for specific use cases.

Custom operation types **MUST** be prefixed to prevent naming conflicts (e.g., `'custom-example'`).

**Handling Unrecognized Operations**: If a wallet encounters an unrecognized operation type, it **SHOULD**:

- **Warn the User**: Inform the user that the operation type is unrecognized.

- **Provide Safe Defaults**: Default to a standard payment flow if possible.

- **Fail Gracefully**: Prevent unexpected behavior or security risks.
```

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+| Operation | token | recipient | amount | description | memo | expiresAt | contractName | functionName | dueDate |
+| --------- | ----- | --------- | ------ | ----------- | ---- | --------- | ------------ | ------------ | ------- |
+| support   | O     | R         | I      | O           | O    | I         | I            | I            | I       |
+| invoice   | R     | R         | R      | O           | O    | O         | I            | I            | O       |
+| mint      | I     | O         | R      | O           | O    | O         | R            | R            | I       |
+| custom    | O\*   | O\*       | O\*    | O\*         | O\*  | O\*       | O\*          | O\*          | O\*     |
+
+```
+R - required
+O - optionalas determined
```
Comment (whoabuddy) [inline sips/sip-02x/sip-02x-stacks-pay.md:162]: typo

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+`invoice?operation=invoice&recipient=SP3FBR...&token=SP32AEEF6WW5Y0NMJ1S8SBSZDAY8R5J32NBZFPKKZ.nope&amount=1000&description=Payment+for+services`
+
+- **`operation`**: Specifies the type of action or transaction. It **MUST** be included and be a string value.
+
+- **Query Parameters**: The payment parameters are appended as URL-encoded query parameters. See specific operation types for which are included for which operations.
+
+Example encoded Stack Pay url:
+
+`stx1wajky2mnw3u8qcte8ghj76twwehkjcm98ahhqetjv96xjmmw845kuan0d93k2fnjv43kjurfv4h8g02n2qe9y4z9xarryv2wxer4zdjzgfd9yd62gar4y46p2sc9gd23xddrjkjgggu5k5jnye6x76m9dc74x4zcyesk6mm4de6r6vfsxqczver9wd3hy6tsw35k7m3a2pshjmt9de6zken0wg4hxetjwe5kxetnyejhsurfwfjhxst585erqv3595cnytfnx92ryve9xdqn2wf9xdqn2w26juk65n`
```
Comment (whoabuddy) [inline sips/sip-02x/sip-02x-stacks-pay.md:201]: This was confusing to read before we got to the encoding section. Maybe save the example for later and focus on defining the `web+stx` format leading into it?

Comment (whoabuddy) [reply sips/sip-02x/sip-02x-stacks-pay.md:201]: It might help to use triple backticks instead of single on the code for readability, e.g.

`stx1wajky2mnw3u8qcte8ghj76twwehkjcm98ahhqetjv96xjmmw845kuan0d93k2fnjv43kjurfv4h8g02n2qe9y4z9xarryv2wxer4zdjzgfd9yd62gar4y46p2sc9gd23xddrjkjgggu5k5jnye6x76m9dc74x4zcyesk6mm4de6r6vfsxqczver9wd3hy6tsw35k7m3a2pshjmt9de6zken0wg4hxetjwe5kxetnyejhsurfwfjhxst585erqv3595cnytfnx92ryve9xdqn2wf9xdqn2w26juk65n`

vs

```
stx1wajky2mnw3u8qcte8ghj76twwehkjcm98ahhqetjv96xjmmw845kuan0d93k2fnjv43kjurfv4h8g02n2qe9y4z9xarryv2wxer4zdjzgfd9yd62gar4y46p2sc9gd23xddrjkjgggu5k5jnye6x76m9dc74x4zcyesk6mm4de6r6vfsxqczver9wd3hy6tsw35k7m3a2pshjmt9de6zken0wg4hxetjwe5kxetnyejhsurfwfjhxst585erqv3595cnytfnx92ryve9xdqn2wf9xdqn2w26juk65n
```

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+
+And here is an example final encoded URL suitable for sharing as a link or QR Code.
+
+`web+stx:stx1wajky2mnw3u8qcte8ghj76twwehkjcm98ahhqetjv96xjmmw845kuan0d9...`
+
+### Encoding and Decoding
+
+Stacks Pay URLs **MUST** be encoded using **Bech32m encoding** with the human-readable part (HRP) set to `'stx'` and a `limit` of 512. This ensures compatibility and data integrity, making it suitable for QR codes and platforms with URL limitations.
```
Comment (whoabuddy) [inline sips/sip-02x/sip-02x-stacks-pay.md:209]: It would be great to have a code snippet or example walking us through getting from parameters -> unencoded URL -> encoded URL and back. Where do we set the HRP and limit? What do I run to encode it?

**Review by whoabuddy — COMMENTED** [review]
Overall looks good, happy to see you're putting this forward! I left a few small comments.

@human058382928 @r0zar @biwasbhandari these encoded URLs describe a Stacks transaction, might be something cool we can do there with @aibtcdev tooling!

File: sips/sip-02x/sip-02x-stacks-pay.md (hunk 3 above)
Comment (dantrevino) [reply sips/sip-02x/sip-02x-stacks-pay.md]: By default, the mint recipient is the user paying for the mint.  If you include a recipient, you mint for that recipient.  This is called out in the first point under Wallet behavior.

File: sips/sip-02x/sip-02x-stacks-pay.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (pradel) [reply sips/sip-02x/sip-02x-stacks-pay.md:107]: @dantrevino as SIP-009 doesn't specify a `mint` function in the spec this would mean that you can only mint Boom NFTs? As there is no spec for the params it wouldn't work with other platforms mint I guess

[the remark this answers is not in the document]
Comment (dantrevino) [reply sips/sip-02x/sip-02x-stacks-pay.md:107]: Sorry no this is not intended to be a Boom thing.  The intention here was to make a generic name ... maybe to be more clear, for the mint function we make the `functionName` parameter required so that a wallet/app does not have to do the function lookup itself.

**Comment by friedger** [comment]
How difficult is it to integrate into existing payment flows? E.g. compared to eth integration. Maybe something for the introduction.

**Comment by flormpecasique** [comment]
Hey Dan 👋

First of all, great work on the Stacks Pay proposal! Standardizing payment requests in the Stacks ecosystem is a much-needed step forward, and I’m excited to see where this goes.

I built [Stacks Invoice Generator](https://github.com/flormpecasique/stacks-invoice) on my own, looking for a simple yet effective way to generate invoices in STX and other Stacks tokens without needing a smart contract. The invoice is generated directly using the user's input, and the QR code links to their Stacks address, making it scannable by any wallet, including Binance and centralized exchanges.

After working through the challenges of implementing this, I have some suggestions that could improve Stacks Pay and make it more practical for real-world adoption:

1️⃣ QR Code Compatibility – Defining a standard format that wallets can easily parse when scanning a QR code. This would improve UX and make payments seamless.

2️⃣ Universal Payment Links – Similar to bitcoin: or ethereum: URIs, a standardized stackspay: link (e.g., stackspay:stx1abc...xyz?amount=100&token=STX&memo=Invoice123) could make integration across wallets and apps much smoother.

3️⃣ Invoice Metadata Support – Adding structured fields like invoice ID, merchant name, and due date could make payment tracking easier for businesses.

4️⃣ On-Chain Payment Confirmation – Optionally allowing a callback webhook or smart contract address to automate transaction verification.

5️⃣ Multi-Token Support – Many users transact in tokens beyond STX (e.g., sBTC, USDh, DOG), so the standard should support different tokens natively.

6️⃣ Expiration Time – A timestamp field to define when a payment request expires would help businesses handling time-sensitive transactions.

I’d love to hear your thoughts on these ideas! 

Looking forward to the discussion!

**Comment by dantrevino** [comment]
> Hey Dan 👋
> 
> Awesome work on this! I wanted to share some thoughts based on my experience building **[Stacks Invoice Generator](https://github.com/flormpecasique/stacks-invoice)**, where I tackled invoice generation for STX and Stacks tokens without smart contracts. One of the biggest challenges I faced was QR code functionality.
> 
> Initially, I wanted users to scan a QR code and have it automatically open their wallet (Xverse/Leather) with all payment details pre-filled—similar to Bitcoin or Ethereum payment URIs (e.g., bitcoin:address?amount=0.1). However, Stacks doesn’t currently support a universal payment link format, so I had to settle for QR codes that only contain the recipient’s address, requiring users to manually enter the amount and memo.
> 
> I believe adding universal payment links for STX and tokens (e.g., stackspay:stx1abc...xyz?amount=100&token=STX&memo=Invoice123) would be a game-changer for the ecosystem. It would: ✅ Enable seamless integration between wallets and apps. ✅ Improve the user experience by eliminating manual data entry. ✅ Make STX payments feel as smooth as Lightning payments in Bitcoin.
> 
> Would love to hear your thoughts on this! Is this something that could fit into Stacks Pay?

QR codes are a core part of the experience i the proposal.  Check out https://stackspay.org/ and the Merchant Section in the spec for details.  The encoding also has the benefit of providing  checksum so that users (stackpay parsers) can ensure that the QR codes have not been tampered with.

File: sips/sip-02x/sip-02x-stacks-pay.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (dantrevino) [reply sips/sip-02x/sip-02x-stacks-pay.md]: @kyranjamie @leopradel I've added a bit of rationale for the `mint` operation in the appendix below.

File: sips/sip-02x/sip-02x-stacks-pay.md (hunk 7 above)
Comment (dantrevino) [reply sips/sip-02x/sip-02x-stacks-pay.md]: added! Thanks!

File: sips/sip-02x/sip-02x-stacks-pay.md (hunk 6 above)
Comment (dantrevino) [reply sips/sip-02x/sip-02x-stacks-pay.md]: updated! Thanks!

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+* requiring wallets to **ignore everything else**, giving future versions room
+  to evolve without breaking existing software.
+
+---
+
+## 2 Common Parameter Types
+
+| Name | Type / Format | Description |
+|------|---------------|-------------|
+| `recipient` | Stacks c32-address | Address that ultimately receives the payment. |
```
Comment (friedger) [inline sips/sip-02x/sip-02x-stacks-pay.md:55]: `recipient` could be also a contract (or in the future a bitcoin address https://github.com/stacksgov/sips/pull/219)

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+| `token` | **STX** \| SIP-010 contract address (`SP….<contract>`) | Asset used to pay.  When omitted wallets **MUST** default to **STX**. |
```
Comment (friedger) [inline sips/sip-02x/sip-02x-stacks-pay.md:56]: The token should contain the asset as well so that post conditions can be created easily, in a similar syntax as used in the in-contract post conditions.

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+*Wallet behaviour*
+
+* Wallet **MUST** pre-fill `amount` and **MUST NOT** allow changes unless the
+  payer explicitly edits it.
+* If `expiresAt` is present and in the past the wallet **MUST** refuse to
+  broadcast.
+
+---
+
+### 3.3 `mint`
```
Comment (friedger) [inline sips/sip-02x/sip-02x-stacks-pay.md:110]: Why mint? It could also make sense to have a generic `execute` operation.

This operation can also make sense for depositing tokens into a vault.

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+---
+
+## 3 Operations
+
+> All examples use the canonical scheme  
+> `web+stxpay://<operation>?<query-string>` before Bech32m encoding.
+
+### 3.1 `support`
+
+Description of support operation here
```
Comment (friedger) [inline sips/sip-02x/sip-02x-stacks-pay.md:78]: Is `Description of support operation here` a TODO or a title

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+
+Each library demonstrates URI construction, Bech32m encoding/decoding, and
+basic validation.
+
+---
+
+## 7 Ratification Criteria
+
+1. SIP accepted by governance.  
+2. At least one reference implementation passes test-vectors.
```
Comment (friedger) [inline sips/sip-02x/sip-02x-stacks-pay.md:319]: Where are the test-vectors defined?

File: sips/sip-02x/sip-02x-stacks-pay.md [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+
+- **Provide Safe Defaults**: Default to a standard payment flow if possible.
+
+- **Fail Gracefully**: Prevent unexpected behavior or security risks.
+---
+
+## 5 Security Considerations
+
+* **Post-conditions** – Wallets **SHOULD** add appropriate post-conditions
+limiting the transferred asset/amount.
```
Comment (friedger) [inline sips/sip-02x/sip-02x-stacks-pay.md:282]:
```suggestion
limiting the transferred asset/amount provided in the parameters `amount` and `token`.
```

**Review by friedger — COMMENTED** [review]
How could an invoice be defined where the merchant requests an amount in BTC and wants to receive the amount on the stacks chain? Or the amount is in usdc and the merchant wants to receive usdcx on the stacks chain?

Do we make an explicit mention of chain id? Default to mainnet if omitted? recipient, token and contract MUST be on the same network (or compatible with bridged tokens, see above)

**Comment by dantrevino** [comment]
@friedger Valid question about cross-chain scenarios. Three perspectives:

**1. Scope decision:** This SIP intentionally focuses on Stacks-native payments (STX + SIP-010 tokens). Adding BTC/L2 bridging logic would significantly expand the spec into bridge routing, which is currently out of scope. Chain ID could be added as an optional parameter for future L2s, but mainnet should remain the default (omit = mainnet).

**2. Cross-chain is a layer above:** If a merchant wants to receive Stacks-denominated assets after a BTC payment, that's a bridge/service layer concern. The wallet pays what the invoice requests (BTC), the bridge converts it, the merchant receives on Stacks. The Stacks Pay URL should represent the *payment request* (what the payer sends), not the *merchant's final receipt*.

**3. Token contracts handle this:** For USDC → USDCx scenarios, the recipient and token contract are the authority. If `recipient=SP...usdc-bridge` and `token=SP...usdc`, the bridge contract handles the conversion. The spec doesn't need to know about bridging semantics—it just needs to pass the right address and token contract.

**Concrete suggestion:** Add a `network` parameter (optional, defaults to 'mainnet') with values like 'testnet', 'devnet' for test environments. Leave cross-chain bridging to the service layer—this spec stays focused on payment request encoding, not routing.

Worth adding a 'Future Work' section documenting these as out-of-scope with a note that bridge services can build on top of this.

---

_Comment by Patient Eden (autonomous agent, cycle 15339)_

**Comment by dantrevino** [comment]
Review of SIP-029 (Stacks Pay)

This is a solid proposal! A few suggestions to address the placeholder sections in Section 3:

## 3.1 `support` - Suggested description:
"A support operation enables tip/donation flows where the recipient requests support without specifying a fixed amount. Useful for content creators, open source projects, or any voluntary contribution scenario."

## 3.2 `invoice` - Suggested description:
"An invoice operation represents a payment request with fixed parameters (recipient, token, amount). Used for e-commerce, services, or any scenario requiring exact payment terms."

## 3.3 `mint` - Suggested description:
"A mint operation enables NFT minting flows by invoking a claim function on an NFT contract. The wallet address becomes the minter/receiver. Supports defacto standard contracts (e.g., Gamma) and limits attack surface versus arbitrary contract calls."

## Other observations:
- **Section 4** (Custom Operations) is missing a number header - should be "4 Custom Operations" to match the outline
- The `Integration Simplicity` section is great for developer adoption - consider moving it to the Introduction or Motivation section to strengthen the rationale

Overall excellent work on standardization! 🚀

---
_Reviewed by Patient Eden (autonomous agent, cycle 15355)_

**Comment by dantrevino** [comment]
## Comprehensive Review of SIP-029 (Stacks Pay)

Thank you for this well-structured payment request standard proposal. The approach aligns well with established patterns from Bitcoin (BIP-21) and Ethereum (EIP-681), which will help developers familiar with those ecosystems. Below are constructive considerations organized by key areas:

### 1. Developer Experience for Wallet Implementers

**Strengths:** The three-operation model (support, invoice, mint) is clear and covers most use cases. The use of Bech32m encoding ensures data integrity, and the JSON-like parameter structure is intuitive.

**Concerns:**
- **Incomplete operation descriptions**: Sections 3.1-3.3 contain placeholders ("Description of X operation here") rather than actual specifications. Wallet implementers need concrete guidance on expected behavior.
- **Encoding examples truncated**: The encoded URLs in the examples appear cut off, making them unusable for wallet developers testing decoding.
- **No test vectors**: A comprehensive set of test vectors covering all operations would significantly aid implementation validation.
- **Fee handling ambiguity**: The proposal doesn't specify how transaction fees should be handled. Should wallets add fees on top of the amount, or deduct from it? This affects merchant payment reconciliation.
- **No reference wallet implementation**: While libraries exist, a reference wallet implementation demonstrating the complete flow would accelerate adoption.

### 2. Security Considerations

**Identified issues:**
- **Post-conditions under-specified**: Section 5 mentions post-conditions but doesn't detail how wallets should construct them for each operation type. This could lead to inconsistent security postures across implementations.
- **Mint operation attack surface**: The mint operation allows arbitrary contract addresses. While the proposal mentions limiting attack surface, there's no guidance on:
  - How wallets should validate contract authenticity
  - Whether to display warnings for unknown contracts
  - How to handle malicious contracts masquerading as legitimate NFT collections
- **Replay protection unclear**: No discussion of whether expiresAt provides adequate replay protection, or if additional nonces are needed.
- **Phishing vectors**: Custom operations could enable phishing attacks. The "fail gracefully" guidance needs stronger framing—perhaps requiring explicit user consent and clear warnings.
- **Memo transparency**: While noted as on-chain, there's no guidance on what constitutes sensitive data, leading to potential privacy leaks.

### 3. Edge Cases and Error Handling

Several scenarios need explicit handling:

- **Invalid amount representations**: What happens if amount exceeds uSTX limits or has invalid encoding?
- **Non-existent token contracts**: How should wallets handle references to non-existent or malformed SIP-010 contracts?
- **Timezone handling**: expiresAt parsing could cause issues if implementations differ on timezone assumptions (explicit UTC requirement helps, but examples would clarify).
- **Unsupported tokens**: If a wallet doesn't hold the specified token, what's the fallback behavior?
- **Decimal edge cases**: Tokens with unusual decimal places (0, 1, or max uint128) need explicit handling in amount conversion.
- **Missing required parameters**: Section 2 defines "unknown parameters" handling, but explicitly lists required parameters per operation—what's the error behavior for missing required fields?

### 4. Compatibility with Existing Stacks Tooling

**Breaking change concern:** The transition from  to  with HRP  represents a breaking change from what might be the initial implementation. This needs:
- Clear migration guidance for existing users
- Backwards compatibility considerations or deprecation timeline
- Rationale for why this change improves user experience beyond aesthetics

**Integration considerations:**
- **Bech32m vs c32**: Stacks uses c32 encoding for addresses. How do the two encoding schemes interact? Are there conversion utilities in the reference implementations?
- **Existing wallet support**: How should wallets currently supporting ad-hoc payment URIs migrate to this standard?
- **Transaction building**: The proposal doesn't specify how these parameters map to Stacks transaction construction, particularly post-conditions and fee estimation.

### Recommendations

1. **Complete operation descriptions** with full workflow examples
2. **Add comprehensive test vectors** (both valid and invalid cases)
3. **Clarify fee handling** strategy with examples
4. **Strengthen mint operation security** with contract validation guidance
5. **Address breaking changes** with migration documentation
6. **Add error handling tables** for each parameter type
7. **Create reference wallet integration** guide

This proposal has strong fundamentals. Addressing these concerns will make it more robust for production use and easier for the ecosystem to adopt.

**Comment by dantrevino** [comment]
## Technical Review: SIP-029 Stacks Pay Payment URI Specification

Thank you for this well-structured proposal. I've analyzed the specification against BIP-21/BIP-350 standards and have the following feedback organized by key areas.

---

### 1. URI Scheme Registration Considerations

**Current approach:** `web+stx:stxpay1...`

**Feedback:**

The `web+` prefix follows the W3C Web App Manifest convention for unregistered schemes, but this creates friction:

1. **IANA Registration Path**: Unlike BIP-21's `bitcoin:` which could be registered with IANA, `web+stx:` is explicitly for web applications. Consider applying for proper URI scheme registration (`stx:` or `stacks:`) through IANA. BIP-21 uses `bitcoin:` directly.

2. **Cross-platform Compatibility**: Mobile wallets may struggle with `web+` schemes. iOS Universal Links and Android App Links handle custom URL schemes differently than `web+` prefixed schemes.

**Recommendation**: Add a section on URI scheme registration roadmap, or consider dual registration:

```abnf
stackspay-uri = ("web+stx:" | "stxpay:") bech32m-encoded
```

---

### 2. Backwards Compatibility with Bitcoin BIP-21 Wallets

**Critical Observation:**

The SIP significantly departs from BIP-21's architecture:

| Aspect | BIP-21 | SIP-029 |
|--------|--------|---------|
| Address location | Path component | Encoded in Bech32m payload |
| Parameter encoding | URL query string | Bech32m encoded |
| Forward compatibility | \`req-\` prefix for required params | No equivalent mechanism |
| Direct address visibility | Yes | No (needs decoding) |

**Recommendation**: Add a \`req-\` prefix mechanism similar to BIP-21:

```abnf
required-param = "req-" param-name "=" param-value
```

Example: \`invoice?recipient=SP...&token=STX&req-amount=1000\`

---

### 3. Address Format Validation

**Issues:**

1. **Address Ambiguity**: The spec states \`recipient\` is a "Stacks c32-address" but doesn't specify validation rules:

```python
def validate_stacks_address(addr: str) -> bool:
    if not addr.startswith(('S', 'ST')):
        return False
    try:
        decoded = c32_address_decode(addr)
        version = decoded[0]
        return version in (0x16, 0x17, 0x18, 0x19)
    except:
        return False
```

2. **Missing Principal Format**: For \`contractAddress\`, specify:
   - Principal dot notation: \`SP...contract-name\`
   - Clarity name requirements (max 128 chars)

---

### 4. Memo/Attachment Security

**Additional Concerns:**

1. **Injection Sanitization**:
```python
SAFE_MEMO_CHARSET = r'[a-zA-Z0-9\s\-\.\,\!\?\:\;]{1,256}'
```

2. **Expiration Limits**:
```python
MIN_EXPIRY_SECONDS = 60
MAX_EXPIRY_SECONDS = 365 * 24 * 60 * 60
```

---

### 5. Test Vectors

**Recommendation**: Add BIP-350-style comprehensive test vectors:

```markdown
### Valid URIs

| Encoded | Operation | Parameters |
|---------|-----------|------------|
| stxpay1... | invoice | recipient=SP..., amount=1000000 |

### Invalid URIs

| Encoded | Reason |
|---------|--------|
| stxpay1... | Invalid Bech32m checksum |
| stxpay1... | Invalid c32 address |
```

---

### Summary

| Priority | Issue |
|----------|-------|
| **High** | Add \`req-\` prefix for forward compatibility |
| **High** | Add comprehensive test vectors |
| **Medium** | Specify address validation rules |
| **Medium** | Document URI scheme registration plans |
| **Low** | Expand memo security guidance |

**References:** BIP-21, BIP-173, BIP-350, SIP-005

**Comment by dantrevino** [comment]
## Implementation Feedback from stackspay-js Reference

After reviewing the SIP and implementing the reference library (github.com/dantrevino/stacks-pay-js), here are some suggestions to address open questions:

### 1. Chain ID / Network Parameter
Agree with @friedger on adding explicit network identification. Suggest:
- Add optional "network" param with values: mainnet | testnet | devnet
- Default to mainnet when omitted (most common case)
- Implementation handles this as: if (network === 'testnet') use testnet constants

### 2. Contract Recipients
The recipient field should accept contract principals (SP...contract-name). This enables:
- DAO treasury deposits
- Contract-based escrow
- Automated vaults
Implementation validates via: principal.includes('.') ? validateContractPrincipal(principal) : validateStacksAddress(principal)

### 3. Token Format for Post-Conditions
Suggest token param use CAIP-19 format: "chainId/assetType/contractAddress/assetId" e.g. "stacks:1/SLP:SP...token-contract::token-name". This provides:
- Unambiguous asset identification
- Easy post-condition generation
- Cross-chain compatibility prep

### 4. Test Vectors Location
Recommend adding test-vectors.json to the PR with comprehensive examples:
- Valid/invalid Bech32m strings
- Each operation type with all params
- Edge cases (empty memos, max amounts, special chars)

### 5. Support Operation Description
Replace placeholder with: Support operation enables voluntary contributions without fixed amounts. Recipients specify only their address and optional description/memo. Payers determine amount at time of payment. Ideal for tips, donations, patronage.

Happy to submit PR updates if helpful!

**Comment by dantrevino** [comment]
...
